kikit and kicad package set

jfly opened the pull request:
kicad: Add support for 3rd party packages

kikit (init at 1.3.0)
kicadAddons.kikit (init at 1.3.0)
kicadAddons.kikit-library (init at 1.3.0)
python310Packages.euclid3 (init at 0.01)
python310Packages.pcbnew-transition (init at 0.3.4)
python310Packages.pybars3 (init at 0.9.7)
python310Packages.pymeta3 (init at 0.5.1)
python311Packages.euclid3 (init at 0.01)
python311Packages.pcbnew-transition (init at 0.3.4)
python311Packages.pybars3 (init at 0.9.7)
python311Packages.pymeta3 (init at 0.5.1)

Co-authored-by: Rohit <rohitsutradhar311@gmail.com>
Co-authored-by: Jeremy Fleischman <jeremyfleischman@gmail.com>
Co-authored-by: Shahar "Dawn" Or <mightyiampresence@gmail.com>
Co-authored-by: Alejandro Sanchez Medina <alejandrosanchzmedina@gmail.com>
Co-authored-by: Matúš Ferech <matus.ferech@gmail.com>


## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

alejandrosame commented:
@ofborg build kicadPackages.kikit kicadPackages.kikit-library python310Packages.euclid3 python310Packages.kikit python310Packages.pcbnew-transition python310Packages.pybars3 python310Packages.pymeta3 python311Packages.euclid3 python311Packages.kikit python311Packages.pcbnew-transition python311Packages.pybars3 python311Packages.pymeta3

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/make-a-derivation-for-a-kicad-plugin/7317/19

jfly committed:
Add jfly to maintainers list

Co-authored-by: Rohit <rohitsutradhar311@gmail.com>
Co-authored-by: Jeremy Fleischman <jeremyfleischman@gmail.com>
Co-authored-by: Shahar "Dawn" Or <mightyiampresence@gmail.com>
Co-authored-by: Alejandro Sanchez Medina <alejandrosanchzmedina@gmail.com>
Co-authored-by: Matúš Ferech <matus.ferech@gmail.com>

jfly committed:
Add matusf to maintainers list

Co-authored-by: Rohit <rohitsutradhar311@gmail.com>
Co-authored-by: Jeremy Fleischman <jeremyfleischman@gmail.com>
Co-authored-by: Shahar "Dawn" Or <mightyiampresence@gmail.com>
Co-authored-by: Alejandro Sanchez Medina <alejandrosanchzmedina@gmail.com>
Co-authored-by: Matúš Ferech <matus.ferech@gmail.com>

jfly committed:
python310Packages.euclid3,python311Packages.euclid3: init at 0.01

Co-authored-by: Rohit <rohitsutradhar311@gmail.com>
Co-authored-by: Jeremy Fleischman <jeremyfleischman@gmail.com>
Co-authored-by: Shahar "Dawn" Or <mightyiampresence@gmail.com>
Co-authored-by: Alejandro Sanchez Medina <alejandrosanchzmedina@gmail.com>
Co-authored-by: Matúš Ferech <matus.ferech@gmail.com>

jfly committed:
python310Packages.pcbnew-transition,python311Packages.pcbnew-transition: init at 0.3.4

Co-authored-by: Rohit <rohitsutradhar311@gmail.com>
Co-authored-by: Jeremy Fleischman <jeremyfleischman@gmail.com>
Co-authored-by: Shahar "Dawn" Or <mightyiampresence@gmail.com>
Co-authored-by: Alejandro Sanchez Medina <alejandrosanchzmedina@gmail.com>
Co-authored-by: Matúš Ferech <matus.ferech@gmail.com>

jfly committed:
python310Packages.pybars3,python311Packages.pybars3: init at 0.9.7

Co-authored-by: Rohit <rohitsutradhar311@gmail.com>
Co-authored-by: Jeremy Fleischman <jeremyfleischman@gmail.com>
Co-authored-by: Shahar "Dawn" Or <mightyiampresence@gmail.com>
Co-authored-by: Alejandro Sanchez Medina <alejandrosanchzmedina@gmail.com>
Co-authored-by: Matúš Ferech <matus.ferech@gmail.com>

jfly committed:
python310Packages.pymeta3,python311Packages.pymeta3: init at 0.5.1

Co-authored-by: Rohit <rohitsutradhar311@gmail.com>
Co-authored-by: Jeremy Fleischman <jeremyfleischman@gmail.com>
Co-authored-by: Shahar "Dawn" Or <mightyiampresence@gmail.com>
Co-authored-by: Alejandro Sanchez Medina <alejandrosanchzmedina@gmail.com>
Co-authored-by: Matúš Ferech <matus.ferech@gmail.com>

jfly committed:
kikit: init at 1.3.0

Co-authored-by: Rohit <rohitsutradhar311@gmail.com>
Co-authored-by: Jeremy Fleischman <jeremyfleischman@gmail.com>
Co-authored-by: Shahar "Dawn" Or <mightyiampresence@gmail.com>
Co-authored-by: Alejandro Sanchez Medina <alejandrosanchzmedina@gmail.com>
Co-authored-by: Matúš Ferech <matus.ferech@gmail.com>

matusf commented:
@evils hey friendly reminder. We've just unbroken installing packages via PCM.

evils commented:
nice

i've been a bit busier than expected, and don't immediately foresee a moment i'll have time

don't hesitate to poke me again in a week if i haven't gotten around to it then

SuperSandro2000 commented on pkgs/applications/science/electronics/kicad/base.nix:
```diff
@@ -68,7 +68,8 @@ stdenv.mkDerivation rec {
   patches = [
     # upstream issue 12941 (attempted to upstream, but appreciably unacceptable)
-    ./writable.patch
+    ./patches/writable.patch
+    ./patches/runtime_stock_data_path.patch
```
```suggestion
    ./writable.patch
    ./runtime_stock_data_path.patch
```
we don't need to move the files into a subdirectory

SuperSandro2000 commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -11,13 +13,16 @@
 , libngspice
 , withScripting ? true
 , python3
+, thirdPartyPackages ? [ ]
```
```suggestion
, extraPythonPackages ? [ ]
```
or something similar to this

jfly replied:
We actually chose the phrase "third party" intentionally. It seems to be a term from kicad that refers to "footprints", "plugins", "resources", or "symbols". Kicad's Plugin and Content Manager (PCM) installs stuff to `~/.local/share/kicad/7.0/3rdparty/`, and the environment variable to configure that location is `KICAD7_3RD_PARTY`.

SuperSandro2000 commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -106,13 +111,42 @@ let
+
+  thirdPartyPackagesJoined =
+    let
+      thirdParty = "$out/share/kicad/scripting";
```
We could just duplicate this to reduce the amount of variables

jfly replied:
Marking resolved as this is done!

SuperSandro2000 commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -106,13 +111,42 @@ let
+      mkdir $out
+
+      for pkg in $thirdPartyPackagesDrvs; do
+        ${lib.getBin unzip}/bin/unzip $pkg/package.zip -d unpacked
```
unzip should be added to nativeBuildInputs instead

jfly replied:
Marking resolved as we did this.

SuperSandro2000 commented on pkgs/applications/science/electronics/kicad/packages/kikit.nix:
```diff
@@ -0,0 +1,45 @@
+}:
+let
+  python = python3.withPackages (ps: with ps; [ click ]);
+  kikit = python3.pkgs.kikit;
```
This should probably live in the python env

jfly replied:
Are you referring to the python env defined on line 9? If so, that wouldn't work: that python env is only used at build time to run some build scripts invoked by the Makefile in the kikit repo.

We need this kikit python dependency while kicad runs. kicad defines its own python interpreter that eventually imports the python code we're packaging here. That interpreter is defined in the kicad derivation.

jfly replied:
Marking resolved. Please reopen if you have any concerns.

SuperSandro2000 commented on pkgs/applications/science/electronics/kicad/packages/kikit.nix:
```diff
@@ -0,0 +1,45 @@
+  inherit (kikit) src version;
+
+  nativeBuildInputs = [ python bc zip ];
+  propagatedBuildInputs = [ kikit ];
```
then we can avoid this which I think only works by coincidence

r0hit05 replied:
We are pretty sure this is necessary for the KiCad python interpreter to be able to import KiKit at runtime. See our comment above for more information.

jfly replied:
Marking resolved. Please reopen if you have any questions.

SuperSandro2000 commented on pkgs/development/python-modules/euclid3/default.nix:
```diff
@@ -0,0 +1,25 @@
+
+  meta = with lib; {
+    description = "2D and 3D vector, matrix, quaternion and geometry module.";
+    homepage = "http://code.google.com/p/pyeuclid/";
```
that site returns a 500

matusf replied:
Maybe temporary outage? Here is what I see:
```
> curl -iL http://code.google.com/p/pyeuclid/
HTTP/1.1 301 Moved Permanently
Content-Type: text/html; charset=utf-8
Location: https://code.google.com/archive/p/pyeuclid
X-Cloud-Trace-Context: 01e9eeda7ec89b111113a686f0e3a287
Date: Mon, 28 Aug 2023 06:22:42 GMT
Server: Google Frontend
Content-Length: 291

HTTP/2 200 
date: Mon, 28 Aug 2023 06:22:42 GMT
expires: Mon, 28 Aug 2023 06:32:42 GMT
cache-control: public, max-age=600
etag: "eunJCg"
x-cloud-trace-context: 798e5f2709bfceacaa407394ef4bef2b
content-type: text/html
server: Google Frontend
alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
```

jfly replied:
Marking this resolved. Please reopen if this is still giving you a 500 somehow.

SuperSandro2000 commented on pkgs/development/python-modules/euclid3/default.nix:
```diff
@@ -0,0 +1,25 @@
+    inherit pname version;
+    hash = "sha256-JbgnpXrb/Zo/qGJeQ6vD6Qf2HeYiND5+U4SC75tG/Qs=";
+  };
+
```
Are there tests to run?

matusf replied:
No, there aren't any tests.

jfly replied:
Marking this resolved. Please reopen if you have a better idea for what to do here!

SuperSandro2000 commented on pkgs/development/python-modules/kikit/default.nix:
```diff
@@ -0,0 +1,87 @@
+
+  # https://github.com/yaqwsx/KiKit/issues/574
+  # copy-pasted from nixpkgs#8d8e62e74f511160a599471549a98bc9e4f4818d
+  shapely = callPackage ./shapely { };
```
This should be added to pythonPackages instead

mightyiam replied:
shapely is [already in nixpkgs](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/shapely/default.nix). We need a much older version here. Does this address your concern, please?

By the way, upstream is [planning to drop shapely](https://github.com/yaqwsx/KiKit/issues/574#issuecomment-1657787964), if that matters.

jfly replied:
Marking this resolved. Please reopen if you'd like us to do something different here.

SuperSandro2000 replied:
> shapely is [already in nixpkgs](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/shapely/default.nix?rgh-link-date=2023-08-28T06%3A28%3A43Z). We need a much older version here. Does this address your concern, please?

Then kitkit needs to move out of pythonPackages.

SuperSandro2000 commented on pkgs/development/python-modules/kikit/default.nix:
```diff
@@ -0,0 +1,87 @@
+  shapely = callPackage ./shapely { };
+in
+buildPythonPackage rec {
+  pname = lib.toLower properCaseName;
```
```suggestion
  pname = "kikit";
```

jfly replied:
Done!

SuperSandro2000 commented on pkgs/development/python-modules/kikit/default.nix:
```diff
@@ -0,0 +1,87 @@
+
+  src = fetchFromGitHub {
+    owner = "yaqwsx";
+    repo = properCaseName;
```
```suggestion
    repo = "KiKit";
```

SuperSandro2000 commented on pkgs/development/python-modules/kikit/default.nix:
```diff
@@ -0,0 +1,87 @@
+    hash = "sha256-kDTPk/R3eZtm4DjoUV4tSQzjGQ9k8MKQedX4oUXYzeo=";
+  };
+
+  propagatedBuildInputs =
+    [
+      kicad
```
```suggestion
  propagatedBuildInputs = [
    kicad
```

jfly replied:
Done!

SuperSandro2000 commented on pkgs/development/python-modules/kikit/default.nix:
```diff
@@ -0,0 +1,87 @@
+  ];
+
+  nativeCheckInputs = [
+    pytest
```
```suggestion
    pytestCheckHook
```

jfly replied:
Done!

SuperSandro2000 commented on pkgs/development/python-modules/kikit/default.nix:
```diff
@@ -0,0 +1,87 @@
+
+  nativeBuildInputs = [
+    versioneer
+    bc
```
Why do we need bc?

jfly replied:
kikit uses `bc` in its Makefile: https://github.com/yaqwsx/KiKit/blob/63301c22ff36b13cc73536b0d289422d3121e772/Makefile#L58

jfly replied:
Marking this resolved. Please reopen if you have any concerns!

SuperSandro2000 commented on pkgs/development/python-modules/kikit/default.nix:
```diff
@@ -0,0 +1,87 @@
+  ];
+
+  checkPhase = ''
+    runHook preCheck
+    export PATH=$PATH:$out/bin
+    make test
+    runHook postCheck
```
We should run the pytest tests with pytestCheckHook and change the directory in preCheck. The bats tests should probably be run in preCheck or postCheck then.

jfly replied:
Marking resolved as this is done.

SuperSandro2000 commented on pkgs/development/python-modules/kikit/shapely/default.nix:
```diff
@@ -0,0 +1,71 @@
+    (substituteAll {
+      src = ./library-paths.patch;
+      libgeos_c = GEOS_LIBRARY_PATH;
+      libc = lib.optionalString (!stdenv.isDarwin) "${stdenv.cc.libc}/lib/libc${stdenv.hostPlatform.extensions.sharedLibrary}.6";
```
Does .dylib.6 exist?

r0hit05 replied:
We copied this file from an old version of nixpkgs. See [this](https://github.com/NixOS/nixpkgs/pull/249464/files/38a7f42312ad59e7604dc1ed01ebc1acaf59849c#r1306964847).

jfly replied:
Marking this resolved. We didn't write this code, but it seems to work, and we'd rather not touch this old code.

Please reopen if you have a suggestion for how to get our hands on an old version of shapely.

SuperSandro2000 commented on pkgs/development/python-modules/kikit/shapely/default.nix:
```diff
@@ -0,0 +1,71 @@
+  '';
+
+  disabledTests = lib.optionals (stdenv.isDarwin && stdenv.isAarch64) [
+    # FIXME(lf-): these logging tests are broken, which is definitely our
+    # fault. I've tried figuring out the cause and failed.
+    #
+    # It is apparently some sandbox or no-sandbox related thing on macOS only
+    # though.
```
This does not explain the error, so I am not even sure if it is our fault.

jfly replied:
Marking this resolved. We didn't write this code, but it seems to work, and we'd rather not touch this old code.

Please reopen if you have a suggestion for how to get our hands on an old version of shapely.

SuperSandro2000 commented on pkgs/development/python-modules/kikit/shapely/default.nix:
```diff
@@ -0,0 +1,71 @@
+
+  meta = with lib; {
+    description = "Geometric objects, predicates, and operations";
+    homepage = "https://pypi.python.org/pypi/Shapely/";
```
```suggestion
    homepage = "https://pypi.org/project/shapely/";
```

jfly replied:
Marking this resolved. We didn't write this code, but it seems to work, and we'd rather not touch this old code.

Please reopen if you have a suggestion for how to get our hands on an old version of shapely.

SuperSandro2000 commented on pkgs/development/python-modules/kikit/solidpython/default.nix:
```diff
@@ -0,0 +1,85 @@
+    euclid3
+
+    prettytable
+    oldPypng
```
We cannot do this in pythonPackages because it will cause collisions

jfly replied:
@SuperSandro2000, I'm not sure I understand your comment. We did *not* put `oldPypng` (or `solidPython`) in `pythonPackages`. Is the concern here about transitive dependencies (that is, someone using `kikit` (which depends on `solidPython` and this `oldPypng`) could run into issues if they also try to use a newer version of `pypng` as well)?

That does seem problematic, but is there anything we can do about that? In practice, we mostly expect people to use this kikit library inside of the kicad project, so the risk of this conflict may be low.

SuperSandro2000 replied:
hmmm, I see. Then kitkit cannot live like this in pythonPackages and should be moved somewhere close to kicat. We don't have package sets in pythonPackages.

alejandrosame replied:
We removed the module from the python package set and packaged the CLI application.

SuperSandro2000 commented on pkgs/development/python-modules/kikit/solidpython/default.nix:
```diff
@@ -0,0 +1,85 @@
+    "PrettyTable"
+  ];
+
+  buildInputs = [
```
```suggestion
  nativeBuildInputs = [
```

jfly replied:
Done!

SuperSandro2000 commented on pkgs/development/python-modules/kikit/solidpython/default.nix:
```diff
@@ -0,0 +1,85 @@
+    description = "Python interface to the OpenSCAD declarative geometry language";
+    homepage = "https://github.com/SolidCode/SolidPython";
+    changelog = "https://github.com/SolidCode/SolidPython/releases/tag/v${version}";
+    maintainers = with maintainers; [ ];
```
According to our contributing guide this is not allowed

jfly replied:
We thought it would be ok to leave this empty as it's not exposed anywhere. I've added myself as a maintainer, so this should be good to go!

SuperSandro2000 commented on pkgs/development/python-modules/pybars3/default.nix:
```diff
@@ -0,0 +1,38 @@
+  ];
+
+  checkPhase = ''
+    runHook preCheck
+    ${python3.interpreter} tests.py
+    runHook postCheck
```
Isn't this just running the unittests with the tests under tests/? Can't pytest run those?

jfly replied:
Neither pytest, nor `python -m unittest` can discover and run these tests due to an ambiguity in this repo about what `tests` refers to. Here's an example of what happens if we let the unittest tool discover and run these tests:

```
ImportError: 'tests' module incorrectly imported from '/build/pybars3-0.9.7/tests'. Expected '/build/pybars3-0.9.7'. Is this module globally installed?
```

Upstream [runs this `tests.py` file in CI](https://github.com/wbond/pybars3/blob/bd2ce03def75bb777da1aa1b0c3884e04382023e/.github/workflows/ci.yml#L39), so we think it's best to do the same thing ourselves.

jfly replied:
Marking this resolved. Please reopen if you think we should do something different.

SuperSandro2000 commented on pkgs/development/python-modules/pymeta3/default.nix:
```diff
@@ -0,0 +1,32 @@
+    hash = "sha256-GL2jJtmpu/WHv8DuC8loZJZNeLBnKIvPVdTZhoHQW8s=";
+  };
+
+  doCheck = false; # Tests do not support Python3
```
:(

jfly replied:
:(

(marking this resolved just to clean up the PR. please reopen if there's something you'd like us to do here)

SuperSandro2000 replied:
Not a big fan of adding already unmaintained software

SuperSandro2000 commented on pkgs/development/python-modules/pymeta3/default.nix:
```diff
@@ -0,0 +1,32 @@
+, fetchPypi
+, lib
+}:
+let
+  properCaseName = "PyMeta3";
+in
+buildPythonPackage rec {
+  pname = lib.toLower properCaseName;
+  version = "0.5.1";
+  format = "setuptools";
+
+  src = fetchPypi {
+    inherit version;
+    pname = properCaseName;
```
```suggestion
buildPythonPackage rec {
  pname = "pymeta3";
  version = "0.5.1";
  format = "setuptools";

  src = fetchPypi {
    inherit version;
    pname = "PyMeta3";
```

jfly replied:
Done!

jfly commented:
@SuperSandro2000, we've pushed a commit addressing most of your feedback. Would appreciate if you could take another look!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-summer-of-nix-program-goals/30376/12

mightyiam commented:
@evils have a moment to review?

evils commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -106,13 +111,40 @@ let
+    runCommand "thirdPartyPackagesJoined"
+      {
+        inherit thirdPartyPackagesDrvs;
+        nativeBuildInputs = [unzip jq];
```
minor style point, but you seem to otherwise be consistent with it:
single line lists are usually padded with spaces inside the square brackets

jfly replied:
Done!

evils reviewed (commented):
the changes in `applications/science/electronics/kicad` look ok to me
i'm unfamiliar with KiKit and don't know how to use these changes
but it builds and is a fairly clean addition to the KiCad wrapper package (`default.nix`)

the closest thing to a blocking issue would be calling the third party packages `kicadPackages`
(it seems like one could expect the KiCad project's `packages3d` there)

matusf commented on pkgs/top-level/all-packages.nix:
```diff
@@ -39470,6 +39470,8 @@ with pkgs;
     with3d = false;
   };
 
+  kicadPackages = recurseIntoAttrs (callPackage ../applications/science/electronics/kicad/packages {});
```
We intend the `kicadAddons` to be the KiCad addons package set. It's not specific to KiKit. 

Regarding `kicad.addons` vs `kicadAddons` we modeled it after vim/vscode plugins. We don't have a strict preference.

evils replied:
ok, that also addresses the plugin discoverability question

i do prefer `kicadAddons` as i would like to keep `kicad.*` as anything directly from the KiCad project or related to its packaging

jfly replied:
No problem! It seems like this is resolved, please reopen if there's anything more to do here.

mightyiam commented:
@evils we addressed your feedback.

mightyiam commented:
Yo, @evils. Ping.

evils commented:
i'm not sure why you're pinging me
i've already reviewed the PR
and as you've stated (and is clear from the github interface) you've addressed my feedback

maybe i should clarify that i don't have commit access

and as i can't actually knowledgeably test the PR due to being unfamiliar with kikit, i'm hesitant to put an approval on it

also note that the review request is put on by the bot automatically and a fairly common thing to ignore, i'll remove mine

evils commented:
not sure how i managed to remove matusf's review request in removing mine, but i assume they don't need a pending request :)

oh and note that some of the later commits should be squashed
especially the ones about responding to PR feedback
and the ones adding comments and then removing those comments

jfly commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -11,13 +13,16 @@
 , libngspice
 , withScripting ? true
 , python3
+, thirdPartyPackages ? [ ]
```
We've added documentation near the top of this file. Please reopen if you have any further suggestions!

jfly commented on pkgs/development/python-modules/pcbnew-transition/default.nix:
```diff
@@ -0,0 +1,39 @@
+  ];
+
+  meta = with lib; {
+    description = "Library that allows you to support both, KiCAD 5 and KiCAD 6 in your plugins";
```
Updated!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1103

SuperSandro2000 commented:
> not sure how i managed to remove matusf's review request in removing mine, but i assume they don't need a pending request :)

One theory: if you change the reviewers, any person that has you blocked is automatically dimissed

SuperSandro2000 commented on pkgs/applications/science/electronics/kicad/default.nix:
```diff
@@ -1,4 +1,6 @@
 { lib, stdenv
```
```suggestion
{ lib
, stdenv
```

mightyiam replied:
Seems unrelated to me.

matusf replied:
Marking this as resolved. We have not touched this part of code, so we think it should not be part of this PR. We are happy to resolve it in a following PR.

jfly committed:
kicadAddons.kikit,kicadAddons.kikit-library: init at 1.3.0

Create a new kicadAddons package set, and a mechanism to override
kicad to bundle addons.

Co-authored-by: Rohit <rohitsutradhar311@gmail.com>
Co-authored-by: Jeremy Fleischman <jeremyfleischman@gmail.com>
Co-authored-by: Shahar "Dawn" Or <mightyiampresence@gmail.com>
Co-authored-by: Alejandro Sanchez Medina <alejandrosanchzmedina@gmail.com>
Co-authored-by: Matúš Ferech <matus.ferech@gmail.com>

mightyiam reviewed (approved):
I'm part of the mob that produced this PR.

mightyiam commented:
Just rebased on nixpkgs-unstable and tested locally again. Here's how to test locally:

```
nix-build -E 'let pkgs = (import ./. {}); in pkgs.kicad.override {addons=[pkgs.kicadAddons.kikit pkgs.kicadAddons.kikit-library];}'
```

Then create a project and look for the KiKit button in the toolbar. Click it. And if you happen to know how to use it, go ahead. We don't.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2824

mightyiam commented:
I contacted @yaqwsx, the author of KiKit for help with verification.

yaqwsx commented:
Hi, I tried building the package, but I am unsure if I use Nix correctly. I run into two problems:
- when I run `./result/bin/pcbnew` I get `Could not determine wxPython version. Python plugins will not be available.`
- I also struggle to find the `kikit` executable.

Once I build the package, how can I install it to test it properly? I also noticed that this package builds KiKit 1.3.0, however, 1.4.0 is the most recent release.

mightyiam commented:
@yaqwsx thank you for trying! Funny. I don't get that wxPython related error.

The kikit executable on this branch can be obtained this way:

```
nix-build -E 'let pkgs = (import ./. {}); in pkgs.kikit'
```

And then look in `result/`

Regarding the version of KiKit, that is just a matter of bumping, I suppose. I'll try doing that.

mightyiam commented:
@yaqwsx would you mind going on a call with me so that I could debug the wxPython error?

yaqwsx commented:
No problem with the call; I am busy for the rest of the week. What timezone are you in so I can suggest a time slot that would suit us both?

evils commented:
weird, `pcbnew` is a wrapper that explicitly sets `PYTHONPATH` to include our `wxPython`
i'm assuming you're not on NixOS and have an ambient `wxPython` installed, and that that's causing an issue

try `nix-shell --pure -I nixpkgs=/path/to/your/nixpkgs/clone/with/these/changes -p kicad`, this should at least clear env vars that may be pointing to another `wxPython`
(though it's entirely possible `pcbnew` finds `wxPython` in an FHS path, dunno what to do about that)
also try launching `pcbnew` via the `kicad` main window

mightyiam commented:
@yaqwsx would you mind scheduling [here](https://koalendar.com/e/chat-with-dawn)?

yaqwsx commented:
@evils Even when I run KiCAD and open Pcbnew from it I get the same error message. When open the pure nix shell, I no longer get the error, but also the KiCAD doesn't see KiKit backend.

mightyiam commented:
> try `nix-shell --pure -I nixpkgs=/path/to/your/nixpkgs/clone/with/these/changes -p kicad`

That would result in a default kicad. Default kicad comes with no addons. Try this to obtain a pure nix shell with kicad and kikit:


```
nix-shell --pure -p 'let pkgs = (import ./. {}); in pkgs.kicad.override {addons=[pkgs.kicadAddons.kikit pkgs.kicadAddons.kikit-library];}'
```

To be clear, when that succeeds, you'd have a `./result/bin/kicad` that should have kikit included. You should also have a `./result/bin/pcbnew` that should work just fine, because of the magic of Nix.

Let us know, please.

yaqwsx commented:
Ok, with shell specififed as above, it seems to work just from KiCAD & Pcbnew. However, what I cannot get working is to invoke KiKit as a command line utility; the following should work: `kikit --help`. I just get `Command not found`.

matusf commented:
Hi, yes, the last expression introduced kikit as only a kicad addon. You can use the following expresion to add kikit itself as well.

```sh
nix-shell --pure -p '
let pkgs = (import ./. {}); in 
[ 
  (pkgs.kicad.override { addons=[pkgs.kicadAddons.kikit pkgs.kicadAddons.kikit-library]; })
  pkgs.kikit
]'
```

mightyiam commented:
> 
> ```
> nix-shell --pure -p 'let pkgs = (import ./. {}); in pkgs.kicad.override {addons=[pkgs.kicadAddons.kikit pkgs.kicadAddons.kikit-library];}'
> ```
> 
> To be clear, when that succeeds, you'd have a `./result/bin/kicad` that should have kikit included. You should also have a `./result/bin/pcbnew` that should work just fine, because of the magic of Nix.

Sorry, @yaqwsx. I mislead you with the paragraph above. `nix-build` produces a `result` directory. `nix-shell`, however, places you in a shell where the executables of the built packages are in the PATH. So, after running a `nix-shell`, don't execute `./result/bin/kicad`. Instead, execute `kicad`.
